### PR TITLE
[subtitles] Improved align to video condition based on resolution ratio

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

Improved the condition used to choose the "aligmnent to video"
with a more reliable condition that check the resolution ratios

added also a check to always align to video the 4/3 subtitles
this because the DVB subtitles at SD (4/3) resolution can be used on a fullhd video
(and fix the regression)

## Motivation and context

An user has reported a regression https://github.com/xbmc/xbmc/issues/20729#issuecomment-1086860869 on a DVB subtitle after PR #21162

## How has this been tested?

Again tested all samples from our FTP folder:
samples_new/subtitles/PGS/
(see FilesInfo.txt for each file info)

and

samples_new/subtitles/DVB/
where i have add the new test case file:
sample1080p_DVB.at.SD.4_3.ts (fullhd video with SD 4/3 DVB subtitle)

PS. no need to test BD 4k menu because the behaviour is the same of before

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp [hunk 1]
```diff
@@ -191,14 +191,15 @@ COverlayTextureGL::COverlayTextureGL(CDVDOverlayImage* o, CRect& rSource)
...
-    int videoSourceH{static_cast<int>(rSource.Height())};
-    int videoSourceW{static_cast<int>(rSource.Width())};
+    float subRatio{static_cast<float>(o->source_width) / o->source_height};
+    float vidRatio{rSource.Width() / rSource.Height()};
 
-    if ((o->source_height == videoSourceH || videoSourceH % o->source_height == 0) &&
-        (o->source_width == videoSourceW || videoSourceW % o->source_width == 0))
+    // We always consider aligning 4/3 subtitles to the video,
+    // for example SD DVB subtitles (4/3) can be rendered on a fullhd video
+    bool isSubResSquare{subRatio == 1.25f || subRatio == 1.33f};
```
Comment (enen92) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp:199]: Can you move this to a new/private method? Easier to read and to reuse. Also please in the docs of this new private method explain what/how the ratios 1.25 and 1.33 can be obtained

**Comment by enen92** [comment]
pinging @thexai for a runtime test with his UHD bluray menus if possible

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp [hunk 2]
```diff
@@ -219,14 +219,15 @@ COverlayImageDX::COverlayImageDX(CDVDOverlayImage* o, CRect& rSource)
...
+    float subRatio{static_cast<float>(o->source_width) / o->source_height};
+    float vidRatio{rSource.Width() / rSource.Height()};
 
-    if ((o->source_height == videoSourceH || videoSourceH % o->source_height == 0) &&
-        (o->source_width == videoSourceW || videoSourceW % o->source_width == 0))
+    // We always consider aligning 4/3 subtitles to the video,
+    // for example SD DVB subtitles (4/3) can be rendered on a fullhd video
+    bool isSubResSquare{subRatio == 1.25f || subRatio == 1.33f};
+
+    if (subRatio == vidRatio || isSubResSquare)
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp]: This may not even work in all cases: float numbers should not be treated as if they were exact values.

A range should be compared to avoid rounding errors:
`a == b` -----> `abs(a - b) < 0.001`

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererGL.cpp]: done

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h [hunk 3]
```diff
@@ -94,6 +94,19 @@ namespace OVERLAY {
...
+    /*!
+     * \brief Given the resolution ratio determines if it is a 4/3 resolution
+     * \param resRatio The resolution ratio (the results of width / height)
+     * \return True if the ratio refer to a 4/3 resolution, otherwise false
+     */
+    bool IsSquareResolution(float resRatio)
+    {
+      // 1:25 is the ratio of old resolutions like PAL (e.g. 720x576)
+      // 1:33 is the ratio of the most common 4/3 resolutions
+      return resRatio == 1.25f || resRatio == 1.33f;
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h:108]: Why 1.33f and not 1.333f or 1.3333f ?

4/3 != 1.33

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp]: from my test not result rounding errors, the ratio values result always correct
but no problem for me add suggested change

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h (hunk 3 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h:108]: I did not obtain reports with infinite value
but i can add same proposed change

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp]: `abs` but return always 0
https://rextester.com/STGYN71172
whats wrong?

Comment (howie-f) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp]: abs works with integers afaik, so if you pass floats they‘re cut. is it `fabs()` ?

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp]: thanks now seem to works!

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp]: change applied

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h (hunk 3 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h]: change applied

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h [hunk 4]
```diff
@@ -94,6 +95,19 @@ namespace OVERLAY {
...
+    /*!
+     * \brief Given the resolution ratio determines if it is a 4/3 resolution
+     * \param resRatio The resolution ratio (the results of width / height)
+     * \return True if the ratio refer to a 4/3 resolution, otherwise false
+     */
+    bool IsSquareResolution(float resRatio)
+    {
+      // 1:25 is the ratio of old resolutions like PAL (e.g. 720x576)
+      // 1:33 is the ratio of the most common 4/3 resolutions
+      return std::fabs(resRatio - 1.25f) < 0.001f || std::fabs(resRatio - 1.33f) < 0.001f;
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h:109]: Do we want it to return true with 640x480 or 768x576 resolutions?

i.e:  `IsSquareResolution(640.0f / 480.0f)` 

I'm also not sure why only these two AR's are tested. Perhaps it would be better to compare with a larger interval to include possible intermediate AR:

`return (resRatio > 1.22f && resRatio < 1.34f)`

Perhaps this fits more in "IsSquareResolution"

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp [hunk 5]
```diff
@@ -219,14 +219,14 @@ COverlayImageDX::COverlayImageDX(CDVDOverlayImage* o, CRect& rSource)
     m_x = (0.5f * o->width + o->x) / o->source_width;
     m_y = (0.5f * o->height + o->y) / o->source_height;
 
-    int videoSourceH{static_cast<int>(rSource.Height())};
-    int videoSourceW{static_cast<int>(rSource.Width())};
+    float subRatio{static_cast<float>(o->source_width) / o->source_height};
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp:222]: Use of braces seems a bit strange here. Can you change to use `=`? 

And maybe use `const float` since this value is not expected to change once calculated.

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h (hunk 4 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h:109]: the two ratios is because is used for DVB subtitle resolution case, not for all video resolutions
maybe i should have given a different name to method

for the dvb subtitle this is correct, we are talking of a use case of DVB subtitle with SD resolution (the two ratios) on a 16/9 res tv channel (as in the code comment),
if this not match is because customised resolutions have been used which should not be tv channels, but more likely you are playing a bad encoded video file

in short, i can extend the range and keep same method name, will probably should have an effect on weirdly encoded video files
IMO these 4/3 contents with uncommon resolutions should not be frequent

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp (hunk 5 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRendererDX.cpp:222]: use braces because it is initialization not assignment
https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#54-initialization

i will add const

File: xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h (hunk 4 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoRenderers/OverlayRenderer.h]: change applied

**Review by thexai — APPROVED** [review]

**Review by enen92 — APPROVED** [review]
